src/engine/strat_engine/backstore/range_alloc.rs

```diff
@@ -12,182 +15,386 @@ use crate::{
+    // Postcondition: result == (None, None) <=> used.len() == 0
+    fn locate_prev_and_next(
+        &self,
+        value: Sectors,
+    ) -> StratisResult<(Option<Sectors>, Option<Sectors>)> {
+        let mut prev = None;
+        let mut next = None;
+        for &key in self.used.keys() {
+            if value >= key {
+                prev = Some(key);
+            }
+            if value <= key {
+                next = Some(key);
+                break;
             }
-        } else {
-            let err_msg = format!(
-                "elements in range ({}, {}) inexpressible in this format",
-                off, len
-            );
-            return Err(StratisError::Engine(ErrorEnum::Invalid, err_msg));
         }
-        Ok(())
+
+        Ok((prev, next))
     }
 
-    /// Mark ranges previously marked as unused as now used.
-    /// Return an error if ranges overlap with each other or with previously
-    /// inserted ranges.
-    /// TODO: Make this operation atomic.
-    /// TODO: Consider using a different algorithmic that first sorts ranges
-    /// and then merges used and ranges by traversing them in parallel, for
-    /// efficiency.
-    fn insert_ranges(&mut self, ranges: &[(Sectors, Sectors)]) -> StratisResult<()> {
-        for &(off, len) in ranges {
-            self.check_for_overflow(off, len)?;
-
-            let prev = self.used.range(..off).rev().next().map(|(k, v)| (*k, *v));
-
-            let mut contig_prev = None;
-            if let Some((prev_off, prev_len)) = prev {
-                if prev_off + prev_len > off {
-                    let err_msg = format!(
-                        "range starting at {} overlaps previous range ({}, {})",
-                        off, prev_off, prev_len
-                    );
-                    return Err(StratisError::Engine(ErrorEnum::Invalid, err_msg));
-                }
-                if prev_off + prev_len == off {
-                    contig_prev = Some((prev_off, prev_len))
-                }
+    // Return the result of what should be obtained on an insertion. A None at
+    // the start or the end of the tuple indicates a contiguous range, hence
+    // a removal from used if it was previously present.
+    // Precondition: prev.is_some() -> prev in self.used
+    // Precondition: next.is_some() -> next in self.used
+    // Precondition: range.1 != 0
+    // Postcondition: prev.is_none() -> result.0.is_none()
+    // Postcondition: next.is_none() -> result.2.is_none()
+    // Postcondition: prev.is_some() && result.0.is_none() -> result.1.0 == prev
+    // Postcondition: range.0 + range.1 <= Sectors max
+    #[allow(clippy::type_complexity)]
+    fn insertion_result(
+        &self,
+        prev: Option<Sectors>,
+        next: Option<Sectors>,
+        range: &(Sectors, Sectors),
+    ) -> StratisResult<(
+        Option<(Sectors, Sectors)>,
+        (Sectors, Sectors),
+        Option<(Sectors, Sectors)>,
+    )> {
+        let (start, len) = range;
```

Comment:
Can pattern matching make this simpler? An example like:

```rust
fn main() {
    let tuple = &(1, 0);
    let &(rc, rs) = tuple;
    println!("{}", rc + 1);
}
```
seems to compile so `rc` is not a reference as long as `Copy` is implemented on the data type which it is for `Sectors`.

Reply:
Yup, that's better!